catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/ws/JsonRpc.java

```diff
@@ -97,83 +101,127 @@ public void onClose(Session session, int statusCode, String reason) {
 
   @Override
   public void onError(Session session, Throwable ex) {
-    // no action required on error
+    if (ex instanceof WebSocketAuthenticationException) {
+      SecurityLogger.audit("Received WebSockets request for user that is not logged in.");
+      handleMessage(
+          session,
+          ((WebSocketAuthenticationException) ex).getWsMessage(),
+          (message, id) -> error(NOT_LOGGED_IN_ERROR, ex.getMessage()));
+    } else {
+      // no action required
+    }
+  }
+
+  @Override
+  public void onMessage(Session session, String message) throws IOException {
+    handleMessage(session, message, this::callMethod);
   }
 
-  private Object exec(Map msg) {
-    if (!msg.containsKey("id")) {
-      return response(null, invalid("required key `id` missing"));
+  private void handleMessage(
+      Session session, String message, BiFunction<Map, Object, Object> handleFunc) {
+    Object id;
+    Object result;
+
+    try {
+      Map messageMap = parseMessage(message);
+      id = parseId(messageMap);
+      validateJsonRpcVersion(messageMap, id);
+      result = handleFunc.apply(messageMap, id);
+    } catch (JsonRpcException exception) {
+      id = exception.messageId;
+      result = exception.error;
     }
 
-    Object id = msg.get("id");
+    session.getRemote().sendStringByFuture(GSON.toJson(response(id, result)));
+  }
+
+  private Map parseMessage(String message) throws JsonRpcException {
+    Map parsed;
+
+    try {
+      parsed = GSON.fromJson(message, Map.class);
+    } catch (RuntimeException ex) {
+      throw new JsonRpcException(null, error(PARSE_ERROR, "Parse error", message));
+    }
+
+    return parsed;
```

Comment:
Not checking that `parsed` is an instance of `Map` because the GSON call used here guarantees it.